### PR TITLE
Features design patterns

### PR DESCRIPTION
**Pull request by Vlad-Magdysh**
•	Created a possibility to accept language identifier
•	Added a possibility to format answers and easily add new formatters, without changing the main logic.
•	Added a possibility to collect configurations of the server from different sources. 
•	Modified C++ client to add a possibility output answer in different destinations, such as MessageBox, file or stdout.

### CODE REVIEW

File: Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp [hunk 1]
```diff
@@ -14,17 +18,61 @@ std::ostream& operator<< (std::ostream& out, const Message& v) {
...
 	}
-	std::cout << std::endl;
 	return out;
 }
 
-int main() {
+class OutputStrategy
+{
+public:
+	virtual void display_answer(const Message& answer, const size_t answer_size) = 0;
```
Comment (Vlad-Magdysh) [inline Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp]: Add a virtual destructor

File: Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp [hunk 2]
```diff
@@ -14,17 +18,61 @@ std::ostream& operator<< (std::ostream& out, const Message& v) {
...
+};
+
+class OutputInFile : public OutputStrategy
+{
+private:
+	std::string file_destination;
+
+public:
+	OutputInFile(const std::string& dest) {
+		file_destination = dest;
```
Comment (Vlad-Magdysh) [inline Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp]: use initialization list in constructor

File: Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp [hunk 3]
```diff
@@ -14,17 +18,61 @@ std::ostream& operator<< (std::ostream& out, const Message& v) {
...
+class OutputStrategy
+{
+public:
+	virtual void display_answer(const Message& answer, const size_t answer_size) = 0;
+};
+
+class OutputInFile : public OutputStrategy
+{
+private:
+	std::string file_destination;
```
Comment (Vlad-Magdysh) [inline Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp]: prefix  m_

File: Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp [hunk 4]
```diff
@@ -14,17 +18,61 @@ std::ostream& operator<< (std::ostream& out, const Message& v) {
...
+
+public:
+	OutputInFile(const std::string& dest) {
+		file_destination = dest;
+	}
+	virtual void display_answer(const Message& answer, const size_t answer_size) final {
```
Comment (Vlad-Magdysh) [inline Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp]: override statement before final

File: Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp [hunk 5]
```diff
@@ -14,17 +18,61 @@ std::ostream& operator<< (std::ostream& out, const Message& v) {
...
+			fs << answer << "\n";
+		}
+	}
+};
+
+class OutputInStdout : public OutputStrategy
+{
+public:
+	virtual void display_answer(const Message& answer, const size_t answer_size) final {
+		std::cout << "RESULT: " << answer;
```
Comment (Vlad-Magdysh) [inline Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp]: \n after the last <<

File: Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp [hunk 6]
```diff
@@ -14,17 +18,61 @@ std::ostream& operator<< (std::ostream& out, const Message& v) {
...
+	}
+};
+
+class OutputInMessageBox : public OutputStrategy
+{
+public:
+	virtual void display_answer(const Message& answer, const size_t answer_size ) final {
+		std::wstring_convert<std::codecvt_utf8<wchar_t>, wchar_t> converter;
```
Comment (Vlad-Magdysh) [inline Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp]: Create a new private function with good function name like convert utf_8_to_utf16

Comment (Vlad-Magdysh) [reply Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp]: + const before wstring

File: Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp [hunk 7]
```diff
@@ -14,17 +18,61 @@ std::ostream& operator<< (std::ostream& out, const Message& v) {
...
+		std::wstring wstr = converter.from_bytes(answer.data(), answer.data() + answer_size);
+		MessageBoxW(HWND_DESKTOP, wstr.c_str(), L"", MB_OK | MB_ICONQUESTION);
+	}
+};
+
+int main(int argc, char* argv[]) {
 	// Setup utf-8 format
 	SetConsoleOutputCP(CP_UTF8);
 	setvbuf(stdout, nullptr, _IOFBF, 1000);
```
Comment (Vlad-Magdysh) [inline Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp]: Remove this line

File: Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp [hunk 8]
```diff
@@ -14,17 +18,61 @@ std::ostream& operator<< (std::ostream& out, const Message& v) {
...
-
 	WSADATA WSAData;
 	const int wsa_startup_status_code = WSAStartup(MAKEWORD(2, 0), &WSAData);
```
Comment (Vlad-Magdysh) [inline Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp]: Use RAII for Startup and Close

File: Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp [hunk 9]
```diff
@@ -35,6 +83,39 @@ int main() {
 	const char* IP = "127.0.0.1";
 	const int PORT = 5555;
 
+	OutputStrategy* output_object = NULL;
```
Comment (Vlad-Magdysh) [inline Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp]: unique_ptr

File: Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp [hunk 10]
```diff
@@ -35,6 +83,39 @@ int main() {
...
+		{
+			output_object = new OutputInStdout();
+		}
+	}
+	else
+	{
+		output_object = new OutputInStdout();
+	}
+
+	std::cout << "Client is running with configured" << typeid(output_object).name() << std::endl;
```
Comment (Vlad-Magdysh) [inline Translator_client_c_plus_plus/Translator_client_c_plus_plus/main.cpp]: Remove

File: Translator_server/main.py [hunk 11]
```diff
@@ -1,56 +1,26 @@
...
-    logger.info(f"Server mode selected: {client_socket_handler.__name__}")
-
-    with SocketManager(ip=IP, port=PORT, listen=arguments.listen) as connections_handler:
+    """
+    Reading Configuration order 
+    - Command line
+    - Local config file
+    - Environment vars
+    """
+    client_socket_handler, socket_manager = FirstConfigHandler.handle()
```
Comment (Vlad-Magdysh) [inline Translator_server/main.py]: move to the main() function

File: Translator_server/handlers.py [hunk 12]
```diff
@@ -1,71 +1,127 @@
 import socket
```
Comment (Vlad-Magdysh) [inline Translator_server/handlers.py]: Rename to request_handler.py

File: Translator_server/handlers.py [hunk 13]
```diff
@@ -1,71 +1,127 @@
...
-        while True:
-            data = None
-            try:
-                data = client_socket.recv(MESSAGE_SIZE)
-            except ConnectionResetError:
-                logger.warning("Client dropped the existing connection")
-            if not data:
-                break
+class RequestParser:
+    """
```
Comment (Vlad-Magdysh) [inline Translator_server/handlers.py]: Move to the file request_parsers.py

File: Translator_server/handlers.py [hunk 14]
```diff
@@ -1,71 +1,127 @@
...
+                    answer = 'EXCEPTION!empty_response'
+                client_socket.sendall(answer.encode("utf-8"))
+                logger.info(f"Response {client_address}:  value = {answer}")
+
+    @property
+    def response_formatter(self) -> Union[AbstractFormatter, None]:
+        return self.__response_formatter
+
+    @response_formatter.setter
+    def response_formatter(self, response_formatter: Union[AbstractFormatter, None]) -> None:
```
Comment (Vlad-Magdysh) [inline Translator_server/handlers.py]: Optional[]

File: Translator_server/handlers.py [hunk 15]
```diff
@@ -1,71 +1,127 @@
...
+            translator (BaseTranslator): translates word from a language to another
+            response_formatter (BaseFormatter): converts an answer from translator in a custom format
+        """
+        # Now the project has only one translator, but I left the opportunity to pass it as parameter to handlers
+        # So in the future it will be easy to add a new translator, read the configs and create a handler with it.
+        self.translator = translator
+        self.response_formatter = response_formatter
+        self.request_parser = request_parser if request_parser is not None else RequestParser()
 
+    def _init_regex(self):
```
Comment (Vlad-Magdysh) [inline Translator_server/handlers.py]: Remove this

File: Translator_server/handlers.py [hunk 16]
```diff
@@ -1,71 +1,127 @@
...
+        """
 
+    def handle_client(self, client_socket: socket, client_address: socket) -> None:
+        """
+        Skeleton of the client processing algorithm
+        """
+        with closing(client_socket):
+            while True:
+                data = None
+                try:
```
Comment (Vlad-Magdysh) [inline Translator_server/handlers.py]: Create a function for this block

File: Translator_server/handlers.py [hunk 17]
```diff
@@ -1,71 +1,127 @@
...
-    :param client_address: Client IP and port, like 127.0.0.1:1234
-    :param client_socket: Client Socket object
-    :return:
-    """
-    return client_processing(client_socket, client_address)
\ No newline at end of file
+                parameters = self.request_parser.parse_request(data.decode("utf-8"))
+                logger.info(
+                    f"Client {client_address} request: "
+                    f"lang = {parameters.get(LANGUAGE_PARAMETER, 'uk')} "
```
Comment (Vlad-Magdysh) [inline Translator_server/handlers.py]: lang and word in separate variables

File: Translator_server/handlers.py [hunk 18]
```diff
@@ -1,71 +1,127 @@
...
+
+                try:
+                    answer = self.translator.translate(word=parameters.get(WORD_PARAMETER, ""),
+                                                       dest=parameters.get(LANGUAGE_PARAMETER, "uk"))
+                except Exception as ex:
+                    logger.error(str(ex))
+                    client_socket.sendall(f"EXCEPTION!{ex}".encode("utf-8"))
+                    continue
+
+                if self.response_formatter is not None:
```
Comment (Vlad-Magdysh) [inline Translator_server/handlers.py]: create function _return_anwer() for this block

File: Translator_server/response_formatters.py [hunk 19]
```diff
@@ -0,0 +1,30 @@
+import json
+from abc import abstractmethod, ABC
+from dicttoxml import dicttoxml
+
+
+class AbstractFormatter(ABC):
+    @abstractmethod
+    def __call__(self, original_data: str) -> str:
```
Comment (Vlad-Magdysh) [inline Translator_server/response_formatters.py]: Rename __call__ to some function

File: Translator_server/response_formatters.py [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+        """
+        pass
+
+class JsonFormatter(AbstractFormatter):
+    def __call__(self, original_data: str) -> str:
+        return json.dumps({"word": original_data}, indent=4)
+
+class XmlFormatter(AbstractFormatter):
+    def __call__(self, original_data: str) -> str:
+        data_to_format = [{"word": original_data}]
```
Comment (Vlad-Magdysh) [inline Translator_server/response_formatters.py]: key "word" in const

File: Translator_client_c_plus_plus/Translator_client_c_plus_plus/WsaDataWrapper.cpp [hunk 21]
```diff
@@ -1,13 +1,15 @@
...
+#include <string> 
+#include <iostream> 
+#include <cstdio> 
 WsaDataWrapper::WsaDataWrapper()
 {
 	const int wsa_startup_status_code = WSAStartup(MAKEWORD(2, 0), &WSAData);
 
 	if (wsa_startup_status_code != 0) {
-		std::cerr << "WSAStartup finished with not zero status code: " << wsa_startup_status_code << std::endl;
+		std::string error_message = "WSAStartup finished with not zero status code: " + std::to_string(wsa_startup_status_code);
```
Comment (Krosik19) [inline Translator_client_c_plus_plus/Translator_client_c_plus_plus/WsaDataWrapper.cpp]: const string

Comment (Krosik19) [reply Translator_client_c_plus_plus/Translator_client_c_plus_plus/WsaDataWrapper.cpp]: code should be translated to error message

Comment (Krosik19) [reply Translator_client_c_plus_plus/Translator_client_c_plus_plus/WsaDataWrapper.cpp]: FormatMessage

File: Translator_client_c_plus_plus/Translator_client_c_plus_plus/WsaDataWrapper.cpp [hunk 22]
```diff
@@ -1,13 +1,15 @@
 #define _WINSOCK_DEPRECATED_NO_WARNINGS
 #include "WsaDataWrapper.h"
-
-
+#include <string> 
+#include <iostream> 
+#include <cstdio>
```
Comment (Krosik19) [inline Translator_client_c_plus_plus/Translator_client_c_plus_plus/WsaDataWrapper.cpp]: what for this header?

File: Translator_client_c_plus_plus/Translator_client_c_plus_plus/WsaDataWrapper.cpp [hunk 23]
```diff
@@ -1,13 +1,15 @@
...
+#include <iostream> 
+#include <cstdio> 
 WsaDataWrapper::WsaDataWrapper()
 {
 	const int wsa_startup_status_code = WSAStartup(MAKEWORD(2, 0), &WSAData);
 
 	if (wsa_startup_status_code != 0) {
-		std::cerr << "WSAStartup finished with not zero status code: " << wsa_startup_status_code << std::endl;
+		std::string error_message = "WSAStartup finished with not zero status code: " + std::to_string(wsa_startup_status_code);
+		throw std::exception(error_message.c_str());
```
Comment (Krosik19) [inline Translator_client_c_plus_plus/Translator_client_c_plus_plus/WsaDataWrapper.cpp]: throw std::exception shouldn't be used

File: Translator_client_c_plus_plus/Translator_client_c_plus_plus/OutputStrategies.h [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+class OutputInStdout : public OutputStrategy
+{
+public:
+	virtual void display_answer(const Message& answer, const size_t answer_size) const override final;
+};
+
+class OutputInMessageBox : public OutputStrategy
+{
+private:
+	std::wstring convert_utf8_to_utf_16(const char* message_start, size_t message_size) const;
```
Comment (Krosik19) [inline Translator_client_c_plus_plus/Translator_client_c_plus_plus/OutputStrategies.h]: should be static or not inside this class

File: Translator_server/server_builder.py [hunk 25]
```diff
@@ -48,25 +48,31 @@ def __setattr__(self, name, value):
...
+    def __init__(self, _next_handler=None):
+        self._next_handler = _next_handler
+
     @abstractmethod
-    def handle(config_object: ConfigurationObject) -> None:
+    def handle(self, config_object: ConfigurationObject) -> None:
         raise NotImplementedError()
 
+    def set_next(self, handler):
+        self._next_handler = handler
```
Comment (Krosik19) [inline Translator_server/server_builder.py]: remove function

File: Translator_server/server_builder.py [hunk 26]
```diff
@@ -48,25 +48,31 @@ def __setattr__(self, name, value):
...
 class EnvVariablesHandler(BaseConfigHandler):
-    @staticmethod
-    def handle(config_object: ConfigurationObject) -> None:
+    def handle(self, config_object: ConfigurationObject) -> None:
         config_object.server_mode = os.getenv("SERVER_MODE", DEFAULT)
         config_object.response_formatter = os.getenv("SERVER_RESPONSE_FORMAT", None)
         config_object.ip = os.getenv("SERVER_IP", "127.0.0.1")
         config_object.port = os.getenv("SERVER_PORT", 5555)
         config_object.listen = os.getenv("SERVER_LISTEN", 4)
+        if self._next_handler is not None:
```
Comment (Krosik19) [inline Translator_server/server_builder.py]: to base class

**Pull request merged**